presto-verifier/src/main/java/com/facebook/presto/verifier/framework/VerificationContext.java

```diff
@@ -0,0 +1,38 @@
+package com.facebook.presto.verifier.framework;
+
+import com.facebook.presto.verifier.framework.QueryOrigin.QueryGroup;
+import com.facebook.presto.verifier.framework.QueryOrigin.QueryStage;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.LinkedListMultimap;
+import com.google.common.collect.ListMultimap;
+
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import static com.facebook.presto.verifier.framework.QueryOrigin.QueryGroup.CONTROL;
+import static com.facebook.presto.verifier.framework.QueryOrigin.QueryGroup.TEST;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+public class VerificationContext
+{
+    private Map<QueryGroup, ListMultimap<QueryStage, String>> failures = new EnumMap<>(QueryGroup.class);
```

Comment:
This is insufficient information in my opinion. To debug the issues we need to have more information. At the least, I am thinking about having the following:

1) presto query id (if we could launch the query)
2) full stack trace
3) error code (something you already have)

Reply:
I think the purpose of exporting each individual failures including the retried ones is to see if there are systematic reliability issue. I'm ok with adding presto query id so that release oncalls can have more power if they want to go and see individual failures (the ones that failed but retried successfully). This is actually already a step involving human effort and ideally we should minimize release oncall having to do this. Stacktrace can be accessed with the query id so I don't think it would be necessary to duplicate them in order xdb.

Reply:
> This is actually already a step involving significant human effort and ideally we should minimize the need for release oncall to do this. (i.e. we should build dashboard around this allFailures fields to show aggregated failure reasons, etc.)

An aggregated view will be good - but during debugging showing more information is better than less. If it feels noisy, it is easier and quicker to remove information from the UI as compared to pumping it into the tooling and the flowing through the DB to the UI.

> Stacktrace can be accessed with the query id so I don't think it would be necessary to duplicate them in order xdb.

As long as we can fetch this information on the UI, I am fine with whichever way you feel is better. The only case where I think this might not be possible with a query id is when we fail launching the query - in that situation as per my understanding since the query didn't even reach presto, we don't even have a query id.

Reply:
I have now finalized the last commit, please take a look again.